Mathlib/Geometry/Manifold/IntegralCurve.lean

```diff
@@ -0,0 +1,767 @@
+
+/-- If there exists `ε > 0` such that the local integral curve at each point `x : M` is defined at
+  least on an open interval `Ioo (-ε) ε`, then every point on `M` has a global integral
+  curve passing through it.
+
+  See Lemma 9.15, Lee -/
+lemma exists_isIntegralCurve_of_isIntegralCurveOn [BoundarylessManifold I M]
+    {v : (x : M) → TangentSpace I x}
+    (hv : ContMDiff I I.tangent 1 (fun x => (⟨x, v x⟩ : TangentBundle I M)))
+    {ε : ℝ} (hε : 0 < ε) (h : ∀ x : M, ∃ γ : ℝ → M, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-ε) ε))
+    (x : M) : ∃ γ : ℝ → M, γ 0 = x ∧ IsIntegralCurve γ v := by
+  have hnon : Set.Nonempty {a | ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)} := ⟨ε, h x⟩
+  by_cases hbdd : BddAbove {a | ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)}
+  · set asup := sSup {a | ∃ γ, γ 0 = x ∧ IsIntegralCurveOn γ v (Ioo (-a) a)} with hasup
+    -- we will obtain two integral curves, one centred at some `t₀ > 0` with
+    -- `0 ≤ asup - ε < t₀ < asup`; let `t₀ = asup - ε / 2`
+    -- another centred at 0 with domain up to `a ∈ S` with `t₀ < a < asup`
+    obtain ⟨a, ha, hlt⟩ := Real.add_neg_lt_sSup hnon (ε := - (ε / 2))
+      (by rw [neg_lt, neg_zero]; exact half_pos hε)
+    have hale : a ≤ asup := le_csSup hbdd ha
+    rw [mem_setOf] at ha
+    rw [← hasup, ← sub_eq_add_neg] at hlt
+
+    have hεle : ε ≤ asup := le_csSup hbdd (h x)
+
+    -- integral curve defined on `Ioo (-a) a`
+    obtain ⟨γ, h0, hγ⟩ := ha
+    -- integral curve starting at `-(asup - ε / 2)` with radius `ε`
+    obtain ⟨γ1_aux, h1_aux, hγ1_aux⟩ := h (γ (-(asup - ε / 2)))
+    let γ1 := γ1_aux ∘ (· + (asup - ε / 2))
+    have heq1 : γ1 (-(asup - ε / 2)) = γ (-(asup - ε / 2)) := by simp [h1_aux]
+    have hγ1 : IsIntegralCurveOn γ1 v (Ioo (-(asup + ε / 2)) (-(asup - ε))) := by
+      apply (hγ1_aux.comp_add (asup - ε / 2)).mono
+      intro t
+      rw [mem_Ioo, mem_setOf, mem_Ioo]
+      intros
+      constructor <;> linarith
+    -- integral curve starting at `asup - ε / 2` with radius `ε`
+    obtain ⟨γ2_aux, h2_aux, hγ2_aux⟩ := h (γ (asup - ε / 2))
+    let γ2 := γ2_aux ∘ (· + -(asup - ε / 2))
+    have heq2 : γ2 (asup - ε / 2) = γ (asup - ε / 2) := by simp [h2_aux]
+    have hγ2 : IsIntegralCurveOn γ2 v (Ioo (asup - ε) (asup + ε / 2)) := by
+      apply (hγ2_aux.comp_add (-(asup - ε / 2))).mono
+      intro t
+      rw [mem_Ioo, mem_setOf, mem_Ioo]
+      rintro ⟨_, _⟩
+      constructor <;> linarith
+
+    -- extend `γ` on the left by `γ1` and on the right by `γ2`
+    set γ_ext : ℝ → M := piecewise (Ioo (-(asup + ε / 2)) a)
+      (piecewise (Ioo (-a) a) γ γ1) γ2 with γ_ext_def
+    have hext : IsIntegralCurveOn γ_ext v (Ioo (-(asup + ε / 2)) (asup + ε / 2)) := by
+      apply (isIntegralCurveOn_piecewise (t₀ := asup - ε / 2) hv _ hγ2
+        (by refine ⟨⟨?_, ?_⟩, ⟨?_, ?_⟩⟩ <;> linarith) _).mono
+      · rintro t ⟨ht1, ht2⟩
+        by_cases hlt : t < a
+        · exact mem_union_left _ ⟨ht1, hlt⟩
```

Comment:
This should also be a separate lemma.

Reply:
I golfed this a little by adding lemmas specialising the sets `s` to the form `Ioo a b` and adding translation lemmas for `sub` mirroring lemmas about `add`.